### DISCUSSION TITLE
TcMenu 4.1.x release discussion

### DISCUSSION THREAD

**Discussion by davetcc** [davetcc/tcMenu#392] [General]
I've started a discussion for TcMenu 4.1.1 which was released today, both library and UI. This release is both for embed control and designer UI along with the associated libraries.

https://github.com/davetcc/tcMenu/releases/tag/4.1.1

Should you run into any release-specific issues, please raise them here and rollback to the prior version.

**Comment by vzahradnik** [comment]
Upgrade from 4.0.0 was smooth for me. But I noticed that TcMenu designer default language is now set to English instead of current system language of my operating system. Manually switching UI languages works as expected.

**Reply by davetcc** [reply]
Please raise a bug for that one.

**Comment by tomelgato** [comment]
Smooth update, TCMenu fix (no X) for long names works

**Comment by davetcc** [comment]
Release 4.1.2 coming out very soon will address three minor issues:

* The dragging of items before submenus
* The new project dialog null pointer exception
* The version CLI command not working properly

**Reply by davetcc** [reply]
This was released yesterday, forgot to update.

**Reply by davetcc** [reply]
I've been told (and then verified) that the disk image does not work on Apple-Silicon based Macs. As luck would have it we have a Macbook Air that is arm64, so we'll get temporary permission to provision it for Notarization so we can make a build.

**Reply by davetcc** [reply]
However, this is more work than you may think, as the certificates are quite fiddly to set up initially, along with that we need to get Xcode tools and Java 20 JDK onto that machine and verify that it can actually build a proper production application.

**Comment by bambo1543** [comment]
Hi Dave, 
with 4.1.1 TitleWidget dosen't redraw when setCurrentState is called. With 4.0.0 it works as expected.

platformio.ini
[env:az-delivery-devkit-v4]
platform = espressif32
board = az-delivery-devkit-v4
framework = arduino
lib_deps = davetcc/tcMenu@4.1.1 // with 4.0.0 TitleWidget redraws
            davetcc/IoAbstraction@4.0.2
            davetcc/tcUnicodeHelper@1.0.0
            davetcc/TaskManagerIO@1.4.1
            bodmer/TFT_eSPI@2.5.31
            bblanchon/ArduinoJson@6.21.3
            robtillaart/ADS1X15@0.3.11
            fbiego/ESP32Time@1.1.0


Cheers,
Andreas

**Reply by davetcc** [reply]
Ack, I'll look into this and fix. Many thanks for pointing out. If I'm honest I had not got around to upgrading my own projects beyond the test boards yet or would have noticed! This was because of a one-off, major change in the rendering layer to fix various bugs that had been reported.

**Reply by davetcc** [reply]
See  #406

**Reply by bambo1543** [reply]
Forgot to mention, that I love tcmenu and it's the best library doing my ESP-Projects.

**Comment by AtmosphEng** [comment]
Hi Dave. Noticed a couple of small things with Designer 4.1.2, embedCONTROL 4.1.1, and PlatformIO build as follows:
Dependency Graph
|-- tcMenu @ 4.1.1
|-- TaskManagerIO @ 1.4.1
|-- IoAbstraction @ 4.0.2
|-- tcUnicodeHelper @ 1.0.0
etc.

1) embedCONTROL - lost previous saved connections and colours. I think new connections are now saved automatically as the Save icon is now missing.
2) A TC menu item with a name that includes an underscore in its middle is viewed / rendered as expected with Designer and on my ESP32-S3 ILI9341 based LCD, but the underscore is removed from the name as seen in the embedCONTROL app on WIndows11.

**Reply by davetcc** [reply]
>embedCONTROL - lost previous saved connections and colours. I think new connections are now saved automatically as the Save icon is now missing.

On this one, it was because I upgraded it to use a database to store all the connections as we got beyond file storage once the ability to override form layout came into play. If you have a lot of these configurations, I can add a command line in 4.2 that converts the XML format into the new format. But I didn't think many were relying on this yet. I think you could use embed control 4.0 with a 4.1 app.

> A TC menu item with a name that includes an underscore in its middle is viewed / rendered as expected with Designer and on my ESP32-S3 ILI9341 based LCD, but the underscore is removed from the name as seen in the embedCONTROL app on WIndows11.

To be honest, I'm not sure I touched any code around that. Is that different from the 4.0 behaviour? I guess it can be fixed to keep underscores, it is nothing but a regex expression that determines what to keep.

**Reply by davetcc** [reply]
See  #406

**Reply by AtmosphEng** [reply]
Thanks. I don't have a lot of previous embedCONTROL configurations and they are easy enough to re-create. No big deal for me.

**Comment by AtmosphEng** [comment]
Hi Dave, I've noticed another couple of small things about embedCONTROL 4.1.1
3. In Color settings: if I select Dark Defaults then Save, the change isn't permanent. If I close and open my project again, it defaults back to Light Defaults.
4. The last version of embedCONTROL that I had was 3.1.0. With that version on Windows 11, I could right-click the embedCONTROL Taskbar icon and choose from a recently used file list. I personally liked this but it doesn't work for me any more on 4.1.1. Also, if I currently run two separate instances of embedCONTROL, then the current default is that both of them open the same most recently used project. So if I want to visually compare two projects, I have to re-open either one or both projects in their respective instances.

**Reply by davetcc** [reply]
I think you're mixing up Embed Control and TcMenu Designer here, TcMenu Designer relies on a packager that changed slightly in that time, if you tell me what platform you're on I can look at that. In terms of the last project loaded at start up, that is simply the last thing you had loaded when it was closed down, it is nothing more than a courtesy that should prevent many from needed to open a project again. 

For embed control, I'll look at the save problem in that issue linked a bit further up.

**Reply by AtmosphEng** [reply]
Oops, sorry. Yes I did mean I was using Designer 4.1.2 on Windows 11 Home for the right-click on the Taskbar icon, which previously showed the most recently used file list for me.

**Reply by davetcc** [reply]
So from 4.2 there's no longer a need to run multiple instances, every time you use File->New, it opens in a new window. Allowing as many projects as needed to be open.

**Comment by AtmosphEng** [comment]
FYI I've been able to successfully use the Designer Code Generator input type 'Control menu with only two buttons' on the hardware factory build of the ESP32-S3 Lilygo T-Display-S3 board. The IO pin assignment is 'Up button' to 14, and 'Down button' to 0. Tried it ok on 'Everything interrupt driven' and 'Polling for keys and encoders'. Thanks.

**Reply by davetcc** [reply]
great news
